config.go

```diff
@@ -16,47 +16,53 @@ var parsersByName = map[string]Parser{
 	"yellowfin": YellowfinLogParser,
 }
 
-type SourceConfig struct {
-	Name     string `json:"name"`
-	Filename string `json:"filename"`
-	Parser   string `json:"parser"`
+type ServiceRegistryConfig struct {
+	Services []struct {
+		Sources []struct {
+			Name     string `json:"name"`
+			Filename string `json:"filename"`
+			Parser   string `json:"parser"`
+		} `json:"sources"`
+	} `json:"services"`
 }
 
-type LogScraperConfig struct {
-	Sources []SourceConfig `json:sources`
-}
-
-func LoadLogScraperConfig(file string) (*LogScraperConfig, error) {
-	if file == "" {
+func LoadServiceRegistryConfig(filename string) (*ServiceRegistryConfig, error) {
+	if filename == "" {
 		return nil, errors.New("Config file cannot be nil")
 	}
 
-	var cfg LogScraperConfig
-	bytes, err := ioutil.ReadFile(file)
+	cfg := &ServiceRegistryConfig{}
+	file, err := os.Open(filename)
 	if err != nil {
 		return nil, err
 	}
 
-	if err := json.Unmarshal(bytes, &cfg); err != nil {
+	defer file.Close()
+	d := json.NewDecoder(file)
+	err = d.Decode(cfg)
+
+	if err != nil {
 		return nil, err
 	}
 
-	if cfg.Sources == nil || len(cfg.Sources) == 0 {
+	if cfg.Services == nil || len(cfg.Services) == 0 {
 		return nil, errors.New("No log sources found in config file")
 	}
 
-	return &cfg, err
+	return cfg, err
 }
 
-func (config *LogScraperConfig) LogSources() ([]*LogSource, []error) {
-	logSources := make([]*LogSource, len(config.Sources))
+func (config *ServiceRegistryConfig) LogSources() ([]*LogSource, []error) {
+	logSources := make([]*LogSource, len(config.Services))
 	errs := make([]error, 0)
 
-	for ind, v := range config.Sources {
-		if _, ok := parsersByName[v.Parser]; ok {
-			logSources[ind] = NewLogSource(v.Name, v.Filename, parsersByName[v.Parser])
-		} else {
-			errs = append(errs, fmt.Errorf("%s has parser %s which cannot be found", v.Name, v.Parser))
+	for ind, v := range config.Services {
+		for _, s := range v.Sources {
+			if _, ok := parsersByName[s.Parser]; ok {
+				logSources[ind] = NewLogSource(s.Name, s.Filename, parsersByName[s.Parser])
```

Comment:
This iteration is causing the logSources array to have null elements when there is no log source for services in the service-registry.json. Use a different index for this, one that you explicitly control and increment every time you find a new log source. This will ensure that the logSources array will never have null elements, and you can remove all the null checks you are doing elsewhere.